src/Features/CSharp/Portable/ConvertLinq/CSharpConvertLinqQueryToForEachProvider.cs

```diff
@@ -0,0 +1,645 @@
+                              initializer,
+                              variableNamePrefix,
+                              out documentUpdate);
+                }
+
+                documentUpdate = default;
+                return false;
+            }
+
+            private bool TryConvertInInvocation(
+                QueryExpressionSyntax source,
+                Func<ExpressionSyntax, ExpressionSyntax, StatementSyntax> leafExpressionCreationMethod,
+                ExpressionSyntax initializer,
+                string variableName,
+                out DocumentUpdate documentUpdate)
+            {
+                var parenthesizedExpression = (ParenthesizedExpressionSyntax)source.Parent;
+                if (parenthesizedExpression.Parent.Parent.Kind() == SyntaxKind.InvocationExpression)
+                {
+                    switch (parenthesizedExpression.Parent.Parent.Parent.Kind())
+                    {
+                        case SyntaxKind.EqualsValueClause:
+                        case SyntaxKind.MultiplyAssignmentExpression:
+                        case SyntaxKind.SimpleAssignmentExpression:
+                            if (TryGetVariableName(parenthesizedExpression.Parent.Parent.Parent, out var identifierExpression))
+                            {
+                                if (TryConvertQueryExpression(source, (ExpressionSyntax expression) => leafExpressionCreationMethod(identifierExpression, expression), out var qe1))
+                                {
+                                    if (TryFindParentAssignment(source, out var parentStatement))
+                                    {
+                                        var updatedParentExpressionStatement = parentStatement.ReplaceNode(
+                                            source.Parent.Parent.Parent,
+                                            initializer);
+
+                                        documentUpdate = new DocumentUpdate(parentStatement, new[] { updatedParentExpressionStatement, qe1 });
+                                        return true;
+                                    }
+                                }
+                            }
+
+                            break;
+                        case SyntaxKind.ReturnStatement:
+                            string freeVariableName = GetFreeSymbolName(variableName, source.GetLocation().SourceSpan.Start);
+                            var identifierName = SyntaxFactory.IdentifierName(freeVariableName);
+                            if (TryConvertQueryExpression(source, (ExpressionSyntax expression) => leafExpressionCreationMethod(identifierName, expression), out var qe2))
+                            {
+                                if (TryFindParentAssignment(source, out var parentStatement))
+                                {
+                                    var variableDeclaration =
```

Comment:
is there a good reason to not use SyntaxGenerator?  it would make a lot of this simpler, and would also potentially allow you to share more logic between C# and VB for this feature.

Reply:
[https://github.com/dotnet/roslyn/issues/25639](https://github.com/dotnet/roslyn/issues/25639)

---
In reply to: [175267717](https://github.com/dotnet/roslyn/pull/25362#discussion_r175267717) [](ancestors = 175267717)